cloud/resource_apikey.go

```diff
@@ -0,0 +1,353 @@
+			},
+			"issued_at": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: descriptions["issued_at"],
+			},
+			"expires_at": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: descriptions["expires_at"],
+			},
+			"private_key": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: descriptions["private_key"],
+			},
+			"key_id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: descriptions["key_id"],
+			},
+			"revoked_at": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: descriptions["revoked_at"],
+			},
+		},
+	}
+}
+
+func resourceApiKeyCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	namespace := d.Get("organization").(string)
+	name := d.Get("name").(string)
+	instanceName := d.Get("instance_name").(string)
+	serviceAccountName := d.Get("service_account_name").(string)
+	description := d.Get("description").(string)
+	expirationTime := d.Get("expiration_time").(string)
+	clientSet, err := getClientSet(getFactoryFromMeta(m))
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("ERROR_INIT_CLIENT_ON_CREATE_API_KEY: %w", err))
+	}
+	ak := &v1alpha1.APIKey{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "APIKey",
+			APIVersion: v1alpha1.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: v1alpha1.APIKeySpec{
+			InstanceName:       instanceName,
+			ServiceAccountName: serviceAccountName,
+		},
+	}
+	r1 := regexp.MustCompile(`^(\d+.)(s|m|h|d)$`)
+	t := time.Now()
+	if expirationTime != "" {
+		if r1.MatchString(expirationTime) {
+			ago, err := str2duration.ParseDuration(expirationTime)
+			if err != nil {
+				return diag.FromErr(fmt.Errorf("ERROR_PARSE_EXPIRATION_TIME: %w", err))
+			}
+			t = t.Add(ago)
+		} else if expirationTime != "No expiration date" {
+			layout := "02/01/2006 15:04:05"
+			t, err = time.Parse(layout, expirationTime)
+			if err != nil {
+				return diag.FromErr(fmt.Errorf("ERROR_PARSE_EXPIRATION_TIME: %w", err))
+			}
+		}
+	} else {
+		defaultExpireTime, err := time.ParseDuration("720h")
+		if err != nil {
+			return diag.FromErr(fmt.Errorf("ERROR_PARSE_DEFAULT_EXPIRATION_TIME: %w", err))
+		}
+		t = t.Add(defaultExpireTime)
+	}
+	if expirationTime != "No expiration date" {
+		ak.Spec.ExpirationTime = &metav1.Time{Time: t}
+	}
+	if description != "" {
+		ak.Spec.Description = description
+	}
+	privateKey, err := util.GenerateEncryptionKey()
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("ERROR_GENERATE_RSA_PRIVATE_KEY: %w", err))
+	}
+	encryptionKey, err := util.ExportPublicKey(privateKey)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("ERROR_EXPORT_PUBLIC_KEY: %w", err))
+	}
+	ak.Spec.EncryptionKey = &v1alpha1.EncryptionKey{
+		PEM: encryptionKey.PEM,
+	}
+	revoke := d.Get("revoke").(bool)
+	ak.Spec.Revoke = revoke
```

Comment:
I think the revoke field value should always be `false` for a creation action

Reply:
It seems from the tests that revoke is allowed to be set to true or false when the apikey is created, also this will help with the test

Reply:
OK

Reply:
Another reason we allow it to be configured at creation phase is because the terraform saves state, and if we always set this field to false during the creation phase and ignore the true field set by the user, this would make the state inconsistent